### PR TITLE
allow an SNI cert to be used to respond for a particular IP

### PR DESCRIPTION
**Pull request by deads2k**
Because the kube-apiserver listens on multiple IPs, it is natural to have different serving certificates for the various IPs (service network, host-network, localhost, etc).  SNI doesn't directly support this from the client-side because when an IP is specified it is unambiguous what server name should be used.  This means that the serverName in the handshake is not set.

By allowing a direct mapping for the certificate to specify an IP and intercepting the clienthello, we can choose a specific certificate that is signed for the IP and will allow the connection to succeed.  This does not introspect all IPs set in the certificate.  We could also add that if we wished

/kind bug
/priority important-soon
@kubernetes/sig-api-machinery-misc 

```release-note
If a serving certificates param specifies a name that is an IP for an SNI certificate, it will have priority for replying to server connections.
```

/hold  I've promised to hold for @liggitt to have comments.

### CODE REVIEW

**Comment by liggitt** [comment]
> or the common name in the certificate to specify an IP

I didn't think common name was supposed to be used in the presence of DNS/IP SANs

**Comment by liggitt** [comment]
I won't be able to get to this until next week at the earliest

**Comment by deads2k** [comment]
> > or the common name in the certificate to specify an IP
> 
> I didn't think common name was supposed to be used in the presence of DNS/IP SANs

Used by clients or used by servers for their configuration?  I claim that this only uses the field as an intent to key for IPs.  Alternatives include:

1. requiring explicit flags to be set
2. reading all IPs

I don't have strong preferences, but I do think we have a need to configure these values.

**Comment by enj** [comment]
/assign

After KubeCon.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+	}, nil)
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	var sniCerts []SNICertKeyContentProvider
+	certSpecs := []testCertSpec{
+		{
+			host:  "172.30.0.1",
+			ips:   []string{"172.30.0.1"},
```
Comment (aojea) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go]: can we add an IPv6 test case?
this test worked for me locally
```
diff --git a/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go b/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go
index 8f0780440e..c8da0eb9df 100644
--- a/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go
+++ b/staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go
@@ -61,6 +61,11 @@ func TestServingCert(t *testing.T) {
                        ips:   []string{"127.0.0.1"},
                        names: []string{"localhost"},
                },
+               {
+                       host:  "2001:abcd:bcda::1",
+                       ips:   []string{"2001:abcd:bcda::1"},
+                       names: []string{"openshiftv6", "openshiftv6.default.svc.cluster.local"},
+               },
        }
 
        for _, certSpec := range certSpecs {
@@ -114,6 +119,7 @@ func TestServingCert(t *testing.T) {
        expectedDefaultCertBytes, _ := defaultCertProvider.CurrentCertKeyContent()
        expectedServiceCertBytes, _ := sniCerts[0].CurrentCertKeyContent()
        expectedLocalhostCertBytes, _ := sniCerts[1].CurrentCertKeyContent()
+       expectedServiceV6CertBytes, _ := sniCerts[2].CurrentCertKeyContent()
 
        tests := []struct {
                name       string
@@ -135,6 +141,11 @@ func TestServingCert(t *testing.T) {
                        serverName: "openshift",
                        expected:   []byte(strings.TrimSpace(string(expectedServiceCertBytes))),
                },
+               {
+                       name:       "service v6 by dns",
+                       serverName: "openshiftv6",
+                       expected:   []byte(strings.TrimSpace(string(expectedServiceV6CertBytes))),
+               },
                {
                        name:       "localhost by dns",
                        serverName: "localhost",
```

**Comment by hickeng** [comment]
@deads2k I had a similar issue and initially wrote something very similar, selecting certs from the SNI set based on IP. This had to change however due to an environment where access to the apiserver was via a NAT load balancer, routed via the overlay, with the consequence that the destination IP did not allow differentiation between external access and access from pods.
I'm noting this here in case this is a scenario of interest to you.

In the end I ended up with the following: https://github.com/kubernetes/kubernetes/compare/v1.16.2...hickeng:cert-filter

I was planning on switching to a more expressive filter syntax, probably a BPF subset, and submitting a PR but have not yet had the time to do so.

This was written before the dynamic certificate work merged - I'm assuming it's better to show you as is rather than after I find time to forward port, possibly after this PR merges. As such it's a diff against v1.16.2.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates_test.go [hunk 2]
```diff
@@ -277,7 +277,7 @@ func generateSelfSignedCertKey(host string, alternateIPs []net.IP, alternateDNS
 	template := x509.Certificate{
 		SerialNumber: big.NewInt(1),
 		Subject: pkix.Name{
-			CommonName: fmt.Sprintf("%s@%d", host, time.Now().Unix()),
+			CommonName: fmt.Sprintf("%s", host),
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates_test.go]: iirc, this was added because different CA certificates with identical serial numbers and common names made some tls verification stacks choke. Was this change required to pass tests?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go (hunk 1 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go]: +1 for ipv6 test

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go [hunk 3]
```diff
@@ -94,7 +95,29 @@ func (c *DynamicServingCertificateController) GetConfigForClient(clientHello *tl
...
+	tlsConfigCopy := tlsConfig.Clone()
+
+	// if the client set SNI information, just use our "normal" SNI flow
+	if len(clientHello.ServerName) > 0 {
+		return tlsConfigCopy, nil
+	}
+
+	// if the client didn't set SNI, then we need to inspect the requested IP so that we can choose
+	// a certificate from our list if we specifically handle that IP.  This can happen when a cert
+	// has a common name that is an IP or when it is specifically mapped by name.
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go]: given the SNI RFC explicitly forbids passing IP addresses to select certs, I'm unsure of the implications of this change

https://tools.ietf.org/html/rfc6066#section-3:

> Literal IPv4 and IPv6 addresses are not permitted in "HostName"

The worst-case scenario I can think of would be:
1. the default serving certificate is one understood/trusted by clients (e.g. signed for DNS name "myapiserver.com")
2. the requested IP address does *not* correspond to the address the client requested (e.g. a network path via a load balancer that makes the requested IP appear to be 127.0.0.1)
3. an additional certificate has a common name (or IP SAN) that matches the requested IP address as seen by the API server

In that scenario, this PR would change the API server's behavior from using the default certificate (which would currently work) to using another configured certificate (which would not work).

My questions for someone who knows (sig-network?) would be:
* can we rely on SNI being passed in the client hello for clients using DNS to speak to the API server?
* can we rely on the requested IP as seen by the apiserver for clients using IP addresses to speak to the API server (I would expect not)

**Comment by liggitt** [comment]
> This sets up those rules, but does not introspect all IPs set in the certificate. We could also add that if we wished

If we're going to try to use IP addresses to select certificates, I would expect the IP SANs to be a much stronger signal than the common name. The common name is not even used to verify IP hostnames in RFC-compliant TLS stacks (e.g. https://github.com/golang/go/blob/master/src/crypto/x509/verify.go#L1012-L1021), so I think selecting a certificate based on the common name to answer an IP-based request would break those clients.

**Comment by liggitt** [comment]
> @deads2k I had a similar issue and initially wrote something very similar, selecting certs from the SNI set based on IP. This had to change however due to an environment where access to the apiserver was via a NAT load balancer, routed via the overlay, with the consequence that the destination IP did not allow differentiation between external access and access from pods.
> I'm noting this here in case this is a scenario of interest to you.

scenarios like that are concerning to me as well

**Comment by deads2k** [comment]
> > This sets up those rules, but does not introspect all IPs set in the certificate. We could also add that if we wished
> 
> If we're going to try to use IP addresses to select certificates, I would expect the IP SANs to be a much stronger signal than the common name. The common name is not even used to verify IP hostnames in RFC-compliant TLS stacks (e.g. https://github.com/golang/go/blob/master/src/crypto/x509/verify.go#L1012-L1021), so I think selecting a certificate based on the common name to answer an IP-based request would break those clients.

As I noted in https://github.com/kubernetes/kubernetes/pull/85308#issuecomment-555009864, I'm happy to make it an explicit configuration choice.  Given the focus of comments here, that seems like a path you'd find acceptable to make the configuration possible even if it isn't easy?

**Comment by liggitt** [comment]
> As I noted in [#85308 (comment)](https://github.com/kubernetes/kubernetes/pull/85308#issuecomment-555009864), I'm happy to make it an explicit configuration choice. Given the focus of comments here, that seems like a path you'd find acceptable to make the configuration possible even if it isn't easy?

Yeah, letting particular deployments opt-in if they know their network topology exposes accurate IP info to the kube-apiserver seems better (since this is done during the handshake, we don't even have the benefit of `X-Forwarded-For`-style headers to convey client request info through a proxy)

**Comment by liggitt** [comment]
Since we already have the ability to indicate explicit hostnames in `--tls-sni-cert-key` options, something like `--tls-sni-cert-key=foo.crt,foo.key:1.2.3.4` could include 1.2.3.4 in a set of IP addresses explicitly opted into this behavior

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go (hunk 1 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go]: added

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go [hunk 4]
```diff
@@ -76,7 +77,10 @@ func getCertificateNames(cert *x509.Certificate) []string {
 	var names []string
 
 	cn := cert.Subject.CommonName
-	if cn == "*" || len(validation.IsDNS1123Subdomain(strings.TrimPrefix(cn, "*."))) == 0 {
+	cnIsIP := net.ParseIP(cn) != nil
+	cnIsValidDomain := cn == "*" || len(validation.IsDNS1123Subdomain(strings.TrimPrefix(cn, "*."))) == 0
+	// don't use the CN if it is a valid IP because our IP serving detection may unexpectedly use it to terminate the connection.
+	if !cnIsIP && cnIsValidDomain {
```
Comment (deads2k) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go]: @liggitt updated logic to prevent the common name from being used if it is an IP

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go (hunk 3 above)
Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go]: @liggitt updated logic to prevent the common name from being used if it is an IP: https://github.com/kubernetes/kubernetes/pull/85308/files#diff-515eb628bf780d6460d8bcc2765dd256R83

**Comment by deads2k** [comment]
Updated to only match on specifically specified sni IPs.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go (hunk 4 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/named_certificates.go]: add to the unit tests to cover this case?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go [hunk 5]
```diff
@@ -0,0 +1,216 @@
...
+	}
+	listener, _, err := createListener("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	apiPort := listener.Addr().String()
+	go func() {
+		t.Logf("listening on %s", listener.Addr().String())
+		listener = tls.NewListener(listener, server.TLSConfig)
+		if err := server.Serve(listener); err != nil {
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go]: does this actually start a TLS server? I expected .ServeTLS(...) (which also requires setting GetCertificate)

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go [hunk 6]
```diff
@@ -0,0 +1,216 @@
...
+		{
+			name:       "localhost by IP",
+			serverName: "127.0.0.1", // this can never actually happen, but let's see
+			expected:   []byte(strings.TrimSpace(string(expectedLocalhostCertBytes))),
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			_, actualDefaultCertBytes, err := cert.GetServingCertificates(apiPort, test.serverName)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go]: just saw the `func GetServingCertificates` impl... looks like that will make the server log `http: TLS handshake error from 127.0.0.1:62829: EOF` on every invocation... mind spawning an issue/follow-up to improve that?

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go [hunk 7]
```diff
@@ -0,0 +1,216 @@
...
+	listener, _, err := createListener("tcp", "127.0.0.1:0")
+	if err != nil {
+		t.Fatal(err)
+	}
+	apiPort := listener.Addr().String()
+	go func() {
+		t.Logf("listening on %s", listener.Addr().String())
+		listener = tls.NewListener(listener, server.TLSConfig)
+		if err := server.Serve(listener); err != nil {
+			t.Error(err)
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go]: also `panic(err)` here, otherwise the test hangs

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go (hunk 6 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go]: nevermind, I misread the utility method and accidentally caused the issue locally by switching the server to `.ServeTLS` and didn't notice you already had a tls listener

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go (hunk 5 above)
Comment (liggitt) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go]: nevermind, you already had a tls listener set up

Comment (deads2k) [reply staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/server_test.go]: > does this actually start a TLS server? I expected .ServeTLS(...) (which also requires setting GetCertificate)

It uses the listener which is configured with TLS configuration. I'll update the line to make it more obvious.

File: staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go (hunk 3 above)
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/dynamiccertificates/tlsconfig.go]: update docs to remove the `common name` aspect and clarify this is a situation where the cluster admin explicitly opted into mapping the incoming IP to a cert

**Comment by liggitt** [comment]
Update the "--tls-sni-cert-key" flag help to describe this new capability (and the implication that it should only be used when the kube-apiserver has visibility to the IP the client requested)

**Comment by liggitt** [comment]
just a couple doc updates, squash, then lgtm

**Comment by liggitt** [comment]
> In the end I ended up with the following: [v1.16.2...hickeng:cert-filter](https://github.com/kubernetes/kubernetes/compare/v1.16.2...hickeng:cert-filter)
> 
> I was planning on switching to a more expressive filter syntax, probably a BPF subset, and submitting a PR but have not yet had the time to do so.

@deads2k did you consider separating the IP-based certs in a way that would allow selecting them by destination IP or by source IP? It's slightly weird to glom them into the SNI flag when we're not actually using SNI here.

**Comment by deads2k** [comment]
> @deads2k did you consider separating the IP-based certs in a way that would allow selecting them by destination IP or by source IP? It's slightly weird to glom them into the SNI flag when we're not actually using SNI here.

It felt pretty natural to describe it via SNI for me, sttts, and our other cluster-admins here.

**Comment by liggitt** [comment]
> It felt pretty natural to describe it via SNI for me, sttts, and our other cluster-admins here.

do you have an idea what you would want to do if you encountered https://github.com/kubernetes/kubernetes/pull/85308#issuecomment-560558108 and needed to limit IP-cert-matching to requests from specific ranges?

**Comment by deads2k** [comment]
> > It felt pretty natural to describe it via SNI for me, sttts, and our other cluster-admins here.
> 
> do you have an idea what you would want to do if you encountered [#85308 (comment)](https://github.com/kubernetes/kubernetes/pull/85308#issuecomment-560558108) and needed to limit IP-cert-matching to requests from specific ranges?

A cluster-admin won't be in worse shape from this today since they simply wouldn't configure the IP listening feature.  I think the logical end state for trying to handle source ranges ends up as a filter chain.  I don't think I'm inclined to go there in the end.

File: staging/src/k8s.io/apiserver/pkg/server/options/serving.go [hunk 8]
```diff
@@ -180,7 +180,9 @@ func (s *SecureServingOptions) AddFlags(fs *pflag.FlagSet) {
 	fs.Var(cliflag.NewNamedCertKeyArray(&s.SNICertKeys), "tls-sni-cert-key", ""+
 		"A pair of x509 certificate and private key file paths, optionally suffixed with a list of "+
 		"domain patterns which are fully qualified domain names, possibly with prefixed wildcard "+
-		"segments. If no domain patterns are provided, the names of the certificate are "+
+		"segments. The domain patterns also allow IP addresses, but IPs should only be use if"+
+		"the apiserver has visibility to the IP address requested by a client."+
```
Comment (liggitt) [inline staging/src/k8s.io/apiserver/pkg/server/options/serving.go]: nit, needs trailing whitespace to join lines correctly
